presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -120,7 +121,9 @@ private boolean exploreNode(int group, Context context)
             }
 
             done = true;
-            for (Rule rule : ruleStore.getPossiblyMatchingRules(node)) {
+            Iterator<Rule> possiblyMatchingRules = ruleStore.getPossiblyMatchingRules(node).iterator();
+            while (possiblyMatchingRules.hasNext()) {
```

Comment:
What is the motivation for this? You still visit all matching rules. Does not materializing list makes such a big difference? Isn't this premature optimization?
Anyway, I am fine with it